### PR TITLE
highlight and search on editor selection (#1633)

### PR DESCRIPTION
**Pull request by jbhoosreddy**
Associated Issue: #1633

### Summary of Changes

* Highlight and search on editor selection by double click
* remove highlight when selection removed
* select previous and next text with key shortcuts.

### CODE REVIEW

File: src/components/Editor.js [hunk 1]
```diff
@@ -382,6 +383,45 @@ const Editor = React.createClass({
...
+    const { query } = this.state;
+    findPrev(ctx, query);
+  },
+
+  traverseResultsNext(e, ctx) {
+    e.stopPropagation();
+    e.preventDefault();
+    const { query } = this.state;
+    findNext(ctx, query);
+  },
```
Comment (jasonLaster) [inline src/components/Editor.js]: hmm, i think these function could receive query and be moved to helpers above the component.

My rationale is that we already have a lot of instance functions :)

File: src/components/Editor.js [hunk 2]
```diff
@@ -382,6 +383,45 @@ const Editor = React.createClass({
...
+    const { searchOverlayMode } = this.state;
+    if (searchOverlayMode) {
+      if (!ctx.cm.somethingSelected()) {
+        removeOverlay(ctx);
+        this.setState({
+          searchOverlayMode: !searchOverlayMode,
+          query: ""
+        });
+      }
+    }
```
Comment (jasonLaster) [inline src/components/Editor.js]: couple things:

1. this seems like a mechanism for removing the search overlay mode. Is that correct?
2. could we also do a serch from here when the user lifts the mouse up
3. the IFs could be collapsed into one `if (searchOverlayMode && !ctx.cm.somethingSelected()) {}`
4. the searchOverlayMode would always be true, so in this case lets just do false... seems more honest

File: src/components/Editor.js [hunk 3]
```diff
@@ -382,6 +383,45 @@ const Editor = React.createClass({
...
+    const query = this.editor.codeMirror.getSelection();
+    this.setState({ searchOverlayMode: true, query });
+    find(ctx, query, true);
+  },
+
+  getInitialState() {
+    return {
+      searchOverlayMode: false
+    };
+  },
```
Comment (jasonLaster) [inline src/components/Editor.js]: shocked that this is the first one

**Comment by nchevobbe** [comment]
This is  really cool !
I find the styling of the outlined element a bit strange, it looks like it cuts-off a bit of them

capture du 2017-01-10 08-41-07

And a bit weird on the selection too (looks like a double border) : 
capture du 2017-01-10 08-42-39

If we could have some padding on this outline it would be great

Another nit, out of scope here, is the selection color in dark mode, which isn't contrasted enough with the text's color (see http://leaverou.github.io/contrast-ratio/#%23E4EEFA-on-%23BDD5FD)

**Comment by jasonLaster** [comment]
@nchevobbe good points. @jbhoosreddy if it's not easy to address, we'll likely want to open a new issue for that.

@bomsy probably has thoughts on class names for styling here

**Comment by bomsy** [comment]
yeah ... we can definitely make the styling better, with paddings and a lil bit of rounded corners. 
The classes responsible are 

https://github.com/devtools-html/debugger.html/blob/master/src/lib/codemirror-mozilla.css#L128-L161

We can do it as another ticket.

**Comment by jasonLaster** [comment]
@jbhoosreddy this PR has gotten really good.

**Comment by jbhoosreddy** [comment]
I looked into improving the CSS for codemirror. I was gonna add it to this
cause it's mostly done. But I can add it in a subsequent PR.

**Pull request merged**